Backend/AlgorithmsManager.py

```diff
@@ -389,6 +389,67 @@ def parse_city_data(self, entry):
         city_data["race_multiple"] = entry[29]
 
         return city_data
+        
+    def get_user_by_username(self, username):
+        print("\nValidating user")
+        try:
+        # def __init__(self, id, ids, uname, signup, loc, protected, folcount, fricount):
+        # self.user_PK = id
+        # self.user_id = ids
+        # self.username = uname
+        # self.signup_date = signup
+        # self.location_id = loc
+        # self.protected = protected
+        # self.followers_count = folcount
+        # self.friends_count = fricount
+            ur = self.twitter.get_user(username)
+            user = DataStructures.User(None, ur.id, ur.screen_name, ur.name, ur.created_at, ur.location,
+                                        ur.protected, ur.followers_count, ur.friends_count)
+            res = []
+            res.append(user.__dict__)
+            print("\nValidation success")
+            return res
+        
+        except:
+            print("Validation fails")
+            return []
+
+    # this function returns a list of retweeters
+    #   args:   username
+    #           num_tweets (optional): how many recent tweets to search through
+    #           num_retweets (optional): how many retweeters per tweet
+    #           sorted (optional): value of 1 sorts the list by number of retweets 
+    #   return: a list of dictionary
+    def get_most_frequent_retweeters(self, username, num_tweets=20, num_retweets=20, sorted=1):
+
+        retweetData = self.twitter.most_recent_retweeters(username, num_tweets=num_tweets, num_retweets=num_retweets)
+        
+        # get the list of all retweeters
+        retweeter_history = []
+        for tweet in retweetData:
+            innerList = tweet['retweeters']
+            for username in innerList:
+                retweeter_history.append(username)
+
+        # count the number of retweets
+        res = []
+        for user_id in retweeter_history:
+            dict = {'username': user_id, 'retweets_count': 1}
```

Comment:
why is 'res' defined as a list of 2-item dictionaries rather than simply as a single dictionary? That is, why not have the result be a dictionary where the keys are usernames and the values are retweet counts?

Reply:
I assume that this list of dictionaries would be returned to the front end as a JSON object. Isn't it easier for the front-end to get each username or retweet_count as attributes that way? Also, we can add further attributes (or items) by not using usernames as keys?

Reply:
Oh, but then you couldn't sort them. Never mind, sorry about the bother.